tests/test_param_routing.py

```diff
@@ -0,0 +1,115 @@
+import inspect
+
+from distutils.version import LooseVersion
+from typing import Any
+from typing import Dict
+
+import numpy as np
+import pytest
+import tensorflow as tf
+
+from tensorflow.keras import Model
+
+from scikeras.wrappers import BaseWrapper
+from scikeras.wrappers import KerasClassifier
+from scikeras.wrappers import KerasRegressor
+
+from .mlp_models import dynamic_classifier
+from .mlp_models import dynamic_regressor
+
+
+def test_routing_basic():
+    """Tests that parameters are routed to the correct place based on static
+    dictionaries.
+    """
+    n, d = 20, 3
+    n_classes = 3
+    X = np.random.uniform(size=(n, d)).astype(float)
+    y = np.random.choice(n_classes, size=n).astype(int)
+
+    def build_fn(*args, **kwargs):
+        assert len(args) == 0, "No *args should be passed to `build_fn`"
+        assert tuple(kwargs.keys()) == (
+            "hidden_layer_sizes",
+            "meta_params",
+            "compile_params",
+        ), "The number and order of **kwargs passed to `build_fn` should be fixed"
+        meta = set(kwargs["meta_params"].keys()) - {"X", "y"}
```

Comment:
I think it's easier to let the users figure out what they want from `X` and `y` instead of trying to guess for them. What's the harm in passing those as parameters?

Reply:
I'm not a fan of implementing unnecessary features until a concrete need comes along.

What use case do you have in mind where having the entire dataset is necessary for model creation? And why doesn't setting a parameter like `model__feature_std=np.std(X)` cover that use case?

> What's the harm in passing those as parameters?

My frame of reference is using SciKeras on a Dask cluster. That means sending large arguments to functions can be rather costly (because of communication). If you'd like, I can detail a specific use case where passing the complete `X` and `y` arrays raise some practical issues.

Reply:
That's fair, and understandable. Previously we were checking if it was an expected parameter, but now we are "forcing" it by passing it in the dict so to speak. I added added/renamed a couple of parameters:
- `cls_type_` → `target_type_`,
- `input_dtype_` → `y_dtype_`
- `y_shape_`
- `X_shape_`
- `X_dtype_`

Reply:
Do you consider this resolved @stsievert ?

Reply:
That's a lot better! Especially with the removal of `X` and `y`.

Style nit: I prefer the `X.shape` and `X.dtype`. I don't see why it's trailing underscores are necessary, especially for `*dtype` and `*shape`.

Reply:
We can't do `X.shape` because these are attributes so it can't have a `.`.

As for the trailing underscore, that's what Scikit-Learn expects: https://scikit-learn.org/stable/developers/develop.html#estimated-attributes
Quote:
> Attributes that have been estimated from the data must always have a name ending with trailing underscore, for example the coefficients of some regression estimator would be stored in a `coef_` attribute